core/src/test/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetServiceTests.java

```diff
@@ -19,103 +19,51 @@
+                final RecoveryTarget recoveryTarget = new RecoveryTarget(replica, null, null, null);
+                assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(0L));
+                recoveryTarget.decRef();
+            }
+            // Last commit is good - use it.
+            final long initDocs = scaledRandomIntBetween(1, 10);
+            {
+                for (int i = 0; i < initDocs; i++) {
+                    indexDoc(replica, "doc", Integer.toString(i));
+                    replica.updateGlobalCheckpointOnReplica(i, "test");
                 }
+                flushShard(replica);
+                final RecoveryTarget recoveryTarget = new RecoveryTarget(replica, null, null, null);
+                assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(initDocs));
+                recoveryTarget.decRef();
+            }
+            // Last commit is not good - use the previous commit
+            {
+                int moreDocs = randomIntBetween(1, 10);
+                for (int i = 0; i < moreDocs; i++) {
+                    indexDoc(replica, "doc", Long.toString(i));
+                    if (rarely()) {
+                        getEngine(replica).getLocalCheckpointTracker().generateSeqNo(); // Create a gap in seqno.
+                    }
+                }
+                flushShard(replica);
+                final RecoveryTarget recoveryTarget = new RecoveryTarget(replica, null, null, null);
+                assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(initDocs));
+                recoveryTarget.decRef();
             }
-
-            final long maxSeqNo = replica.seqNoStats().getMaxSeqNo();
-            final long localCheckpoint = replica.getLocalCheckpoint();
-
-            translogLocation.set(replica.getTranslog().location());
-
-            final Translog translog = replica.getTranslog();
-            final String translogUUID = translog.getTranslogUUID();
-            assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(0L));
-
-            translogLocation.set(writeTranslog(replica.shardId(), translogUUID, translog.currentFileGeneration(), maxSeqNo - 1));
-
-            // commit is good, global checkpoint is at least max *committed* which is NO_OPS_PERFORMED
-            assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(0L));
-
-            replica.flush(new FlushRequest());
-
-            translogLocation.set(replica.getTranslog().location());
-
-            // commit is not good, global checkpoint is below max
-            assertThat(PeerRecoveryTargetService.getStartingSeqNo(recoveryTarget), equalTo(SequenceNumbers.UNASSIGNED_SEQ_NO));
-
-            translogLocation.set(writeTranslog(replica.shardId(), translogUUID, translog.currentFileGeneration(), maxSeqNo));
-
-            // commit is good, global checkpoint is above max
```

Comment:
we lost the extra check that advancing the global checking advanceses the commit usage.